src/compiler/program.ts

```diff
@@ -211,6 +221,77 @@ namespace ts {
+            if (!host.hasChanges) {
+                // host does not support method 'hasChanges'
+                return false;
+            }
+            if (!oldProgram) {
+                return false;
+            }
+            
+            Debug.assert(!oldProgram.structureIsReused);
+
+            // there is an old program, check if we can reuse its structure
+            let oldRootNames = oldProgram.getRootFileNames();
+            if (rootNames.length !== oldRootNames.length) {
+                // different amount of root names - structure cannot be reused
+                return false;
+            }
+            
+            for (let i = 0; i < rootNames.length; i++) {
+                if (oldRootNames[i] !== rootNames[i]) {
+                    // different order of root names - structure cannot be reused
+                    return false;
+                }
+            }
+            
+            // check if program source files has changed in the way that can affect structure of the program
+            let newSourceFiles: SourceFile[] = [];
+            for (let oldSourceFile of oldProgram.getSourceFiles()) {
+                let newSourceFile: SourceFile;
+                if (host.hasChanges(oldSourceFile)) {
+                    newSourceFile = host.getSourceFile(oldSourceFile.fileName, options.target);
+                    if (!newSourceFile) {
+                        return false;
+                    }
+                    
+                    // check tripleslash references
+                    if (!arrayIsEqualTo(oldSourceFile.referencedFiles, newSourceFile.referencedFiles, fileReferenceIsEqualTo)) {
+                        // tripleslash references has changed
+                        return false;
+                    }
+                    
+                    // check imports
+                    collectExternalModuleReferences(newSourceFile);                    
+                    if (!arrayIsEqualTo(oldSourceFile.imports, newSourceFile.imports, moduleNameIsEqualTo)) {
+                        // imports has changed
+                        return false;
+                    }
+                    // pass the cache of module resolutions from the old source file
+                    newSourceFile.resolvedModules = oldSourceFile.resolvedModules;
+                }
+                else {
+                    // file has no changes - use it as is
+                    newSourceFile = oldSourceFile;
+                }
+                
+                // if file has passed all checks it should be safe to reuse it
+                newSourceFiles.push(newSourceFile);
+            }
+            
+            // update fileName -> file mapping
+            for (let file of newSourceFiles) {
+                filesByName.set(file.fileName, file);
+            }
+            
+            files = newSourceFiles;
+            
+            oldProgram.structureIsReused = true;

```

Comment:
what does this give us at this point? we have copied the data, there should not be any dependency, correct?